### PR TITLE
Function to modify key codec

### PR DESCRIPTION
**Pull request by austinabell**
Hey, not sure if this is disallowed for any reason specifically, but it would be very beneficial to define additional types to decode tendermint key implementations from bytes, since it uses a static codec. If this is okay, let me know and I will add documentation.

Context: For Ethermint to switch to using Cosmos' keybase, decoding the keys requires this codec to be updated

* [ ] Referenced an issue explaining the need for the change
* [ ] Updated all relevant documentation in docs
* [ ] Updated all code comments where relevant
* [ ] Wrote tests
* [ ] Updated CHANGELOG_PENDING.md

### CODE REVIEW

**Comment by melekes** [comment]
this needs a CHANGELOG_PENDING.md entry

```
- [crypto] Add `RegisterKeyType` to amino to allow external key types registration (@austinabell)
```

**Comment by melekes** [comment]
Refs https://github.com/tendermint/tendermint/issues/2358

**Comment by austinabell** [comment]
Just to document, I did experiment with creating a mapping from string to objects to be able to keep track of the key types added to be able to be used in the `RegisterAmino(..)` call, but because of how go is compiled, cosmos would just use the base types. This may be a useful feature for someone just building on top of Tendermint and not going through Cosmos, but to not add confusion or unnecessary complexity, I left it out.

File: crypto/encoding/amino/amino.go [hunk 1]
```diff
@@ -60,11 +60,20 @@ func RegisterAmino(cdc *amino.Codec) {
 		secp256k1.PrivKeyAminoName, nil)
 }
 
+// RegisterKeyType registers an external key type to allow decoding it from bytes
+func RegisterKeyType(o interface{}, name string) {
+	cdc.RegisterConcrete(o, name, nil)
+
+	nameTable[reflect.TypeOf(o)] = name
```
Comment (melekes) [inline crypto/encoding/amino/amino.go]: Should we check if an entry with the same name already exists before possibly overriding it?

File: crypto/encoding/amino/amino.go [hunk 2]
```diff
@@ -60,11 +60,20 @@ func RegisterAmino(cdc *amino.Codec) {
 		secp256k1.PrivKeyAminoName, nil)
 }
 
+// RegisterKeyType registers an external key type to allow decoding it from bytes
+func RegisterKeyType(o interface{}, name string) {
```
Comment (melekes) [inline crypto/encoding/amino/amino.go]: Would be great to have a test case for it.

File: crypto/encoding/amino/amino.go (hunk 1 above)
Comment (austinabell) [reply crypto/encoding/amino/amino.go]: Yeah, I'll do that now. Do you want me to use the added types to be used in `RegisterAmino`?

File: crypto/encoding/amino/amino.go (hunk 2 above)
Comment (austinabell) [reply crypto/encoding/amino/amino.go]: Will do

**Comment by austinabell** [comment]
Sorry that took a little longer than expected, test setup was a little finicky and the print tests required me to make sure the codec was reset after the test. Alternatively I could have changed the expected print output but I didn't want to change any existing tests

File: crypto/encoding/amino/amino.go [hunk 3]
```diff
@@ -62,8 +62,11 @@ func RegisterAmino(cdc *amino.Codec) {
 
 // RegisterKeyType registers an external key type to allow decoding it from bytes
 func RegisterKeyType(o interface{}, name string) {
-	cdc.RegisterConcrete(o, name, nil)
+	if _, exists := nameTable[reflect.TypeOf(o)]; exists {
+		return
```
Comment (melekes) [inline crypto/encoding/amino/amino.go]: I think we should panic, not silently return

File: crypto/encoding/amino/encode_test.go [hunk 4]
```diff
@@ -147,3 +149,75 @@ func TestPubkeyAminoName(t *testing.T) {
...
+}
+func (key TestPub) VerifyBytes(msg []byte, sig []byte) bool { return true }
+func (key TestPub) Equals(other crypto.PubKey) bool         { return true }
+
+var (
+	privAminoName = "registerTest/Priv"
+	pubAminoName  = "registerTest/Pub"
+)
+
+func TestRegisterKeyType(t *testing.T) {
```
Comment (melekes) [inline crypto/encoding/amino/encode_test.go]: this is great!

File: crypto/encoding/amino/amino.go (hunk 3 above)
Comment (austinabell) [reply crypto/encoding/amino/amino.go]: Yeah, I can do that, but isn't that basically the same panic that occurs in `RegisterConcrete` when you try to register an existing type?

Comment (melekes) [reply crypto/encoding/amino/amino.go]: Did not know we panic in Register concrete. If that's the case, we can get rid of the extra check. Thank you

File: crypto/encoding/amino/encode_test.go [hunk 5]
```diff
@@ -2,10 +2,12 @@ package cryptoAmino
 
 import (
 	"os"
+	"reflect"
 	"testing"
 
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
+	amino "github.com/tendermint/go-amino"
```
Comment (melekes) [inline crypto/encoding/amino/encode_test.go]:
```suggestion

	amino "github.com/tendermint/go-amino"

```

**Comment by ebuchman** [comment]
Does this actually work in a real case? I thought the codecs are declared privately on a per package basis so it would be hard to get this effect to thread through ?

**Comment by austinabell** [comment]
> Does this actually work in a real case? I thought the codecs are declared privately on a per package basis so it would be hard to get this effect to thread through ?

Does for what I needed it for in Ethermint, can't say for other uses. If defined in the main() function of the cli and daemon binary as such:
```
	tmamino.RegisterKeyType(emintcrypto.PubKeySecp256k1{}, emintcrypto.PubKeyAminoName)
	tmamino.RegisterKeyType(emintcrypto.PrivKeySecp256k1{}, emintcrypto.PrivKeyAminoName)
```

It is able to be overriden for when a public key is decoded through the `..FromBytes` functions using that codec.

I have noticed other interactions haven't threaded through based on how dependencies of dependencies are compiled in go which made other things more annoying, but this PR allowed the Ethermint key type to be decoded from the Cosmos keybase.

**Pull request merged**